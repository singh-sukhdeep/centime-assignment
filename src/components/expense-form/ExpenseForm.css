

.form-wrapper{
    width: 50%;
    gap: 1rem;
    display: flex;
    flex-direction: column;
    padding: 1rem;
}
.form-fields{
    display: flex;
    flex-direction: column;
    gap: 1rem;

}

.form-field{
    display: flex;
    align-items: center;
}

.form-field__label{
    flex: 1;
}

.form-field__input{
    flex: 1;
  
    display: flex;
    flex-direction: column;
    gap: 0.5rem;
   
}

.form-field__input select,input{
    height: 1.5rem;
}

.form-field__input p{
    margin: 0;
    color: red;
}

.form-submit{
    display: flex;
    gap: 0.5rem;
    justify-content: flex-end;
}