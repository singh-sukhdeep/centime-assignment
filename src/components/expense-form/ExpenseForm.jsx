import { useFormik } from "formik";
import { useTranslation } from "react-i18next";
import * as yup from "yup";

import "./ExpenseForm.css";
import { useEffect } from "react";

export default function ExpenseForm({
    fromSelectValues = [],
    editForm,
    onExpenseSubmit,
}) {
    // Translation
    const { t } = useTranslation();

    const validationSchema = yup.object({
        from: yup.string("From is Required").required("From is required"),
        expenseLabel: yup
            .string("Expense Label is string")
            .required("Expense Label is required"),
        amount: yup
            .number("Amount must be number")
            .positive("Amount must be positive Number")
            .required("Amount is Required"),
    });
    const formik = useFormik({
        initialValues: {
            from: "Incomes",
            expenseLabel: "",
            amount: "",
        },
        validationSchema: validationSchema,
        onSubmit: (values) => {
            onExpenseSubmit?.(values);
            formik?.resetForm();
        },
        onReset: (value) => { },
    });

    useEffect(() => {
        if (editForm) {
            formik.setValues(editForm, true);
        }
    }, [editForm]);

    return (
        <form
            className="form-wrapper"
            name="expense-form"
            data-testid="expense-form-test"
            onSubmit={formik.handleSubmit}
            onReset={formik.handleReset}
        >
            <div className="form-fields">
                <div className="form-field">
                    <label className="form-field__label" htmlFor="email">
                        {t("expenseForm.fromLabel")}
                    </label>
                    <div className="form-field__input">
                        <select
                            data-testid="from-test"
                            name="from"
                            id="from"
                            value={formik.values.from}
                            onChange={formik.handleChange}
                        >
                            {fromSelectValues.map((fromSelectValue) => {
                                return (
                                    <option
                                        key={fromSelectValue.value}
                                        value={fromSelectValue.value}
                                    >
                                        {fromSelectValue.label}
                                    </option>
                                );
                            })}
                        </select>
                        {formik.touched.from && Boolean(formik.errors.from) && (
                            <p className="error">{formik.errors.from}</p>
                        )}
                    </div>
                </div>

                <div className="form-field">
                    <label className="form-field__label" htmlFor="expenseLabel">
                        {t("expenseForm.expenseLabel")}
                    </label>
                    <div className="form-field__input">
                        <input
                            data-testid="expenseLabel-test"
                            id="expenseLabel"
                            name="expenseLabel"
                            type="expenseLabel"
                            onChange={formik.handleChange}
                            value={formik.values.expenseLabel}
                        />
                        {formik.touched.expenseLabel &&
                            Boolean(formik.errors.expenseLabel) && (
                                <p className="error">{formik.errors.expenseLabel}</p>
                            )}
                    </div>
                </div>

                <div className="form-field">
                    <label className="form-field__label" htmlFor="amount">
                        {t("expenseForm.amountLabel")}
                    </label>
                    <div className="form-field__input">
                        <input
                            data-testid="amount-test"
                            id="amount"
                            name="amount"
                            type="number"
                            onChange={formik.handleChange}
                            value={formik.values.amount}
                        />
                        {formik.touched.amount && Boolean(formik.errors.amount) && (
                            <p className="error">{formik.errors.amount}</p>
                        )}
                    </div>
                </div>
            </div>

            <div className="form-submit">
                <button data-testid="reset-button-test" type="reset">{t("expenseForm.resetButton")}</button>
                <button data-testid="submit-button-test" type="submit">{t("expenseForm.submitButton")}</button>
            </div>
        </form>
    );
}
