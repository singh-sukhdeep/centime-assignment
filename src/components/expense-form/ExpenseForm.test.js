import React from 'react';
import { render, fireEvent, waitFor,screen } from '@testing-library/react';
import ExpenseForm from './ExpenseForm';

// Mock the useTranslation hook
jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: key => key }),
}));

describe('ExpenseForm', () => {
  it('renders the component with default values', () => {
    render(<ExpenseForm />);

    // console.log(getByLabelText,getByText,'===============');
    expect(screen.getByText('expenseForm.fromLabel')).toBeInTheDocument();
  });

});
