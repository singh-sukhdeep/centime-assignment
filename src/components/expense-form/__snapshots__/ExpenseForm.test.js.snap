// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expense Form snapshot 1`] = `
<form
  className="form-wrapper"
  data-testid="expense-form-test"
  name="expense-form"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="form-fields"
  >
    <div
      className="form-field"
    >
      <label
        className="form-field__label"
        htmlFor="email"
      >
        expenseForm.fromLabel
      </label>
      <div
        className="form-field__input"
      >
        <select
          data-testid="from-test"
          id="from"
          name="from"
          onChange={[Function]}
          value="Incomes"
        >
          <option
            value="Incomes"
          >
            Incomes
          </option>
          <option
            value="Salary"
          >
            Salary
          </option>
        </select>
      </div>
    </div>
    <div
      className="form-field"
    >
      <label
        className="form-field__label"
        htmlFor="expenseLabel"
      >
        expenseForm.expenseLabel
      </label>
      <div
        className="form-field__input"
      >
        <input
          data-testid="expenseLabel-test"
          id="expenseLabel"
          name="expenseLabel"
          onChange={[Function]}
          type="expenseLabel"
          value=""
        />
      </div>
    </div>
    <div
      className="form-field"
    >
      <label
        className="form-field__label"
        htmlFor="amount"
      >
        expenseForm.amountLabel
      </label>
      <div
        className="form-field__input"
      >
        <input
          data-testid="amount-test"
          id="amount"
          name="amount"
          onChange={[Function]}
          type="number"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="form-submit"
  >
    <button
      data-testid="reset-button-test"
      type="reset"
    >
      expenseForm.resetButton
    </button>
    <button
      data-testid="submit-button-test"
      type="submit"
    >
      expenseForm.submitButton
    </button>
  </div>
</form>
`;

exports[`Expense Form snapshot test 1`] = `
<form
  className="form-wrapper"
  data-testid="expense-form-test"
  name="expense-form"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="form-fields"
  >
    <div
      className="form-field"
    >
      <label
        className="form-field__label"
        htmlFor="email"
      >
        expenseForm.fromLabel
      </label>
      <div
        className="form-field__input"
      >
        <select
          data-testid="from-test"
          id="from"
          name="from"
          onChange={[Function]}
          value="Incomes"
        >
          <option
            value="Incomes"
          >
            Incomes
          </option>
          <option
            value="Salary"
          >
            Salary
          </option>
        </select>
      </div>
    </div>
    <div
      className="form-field"
    >
      <label
        className="form-field__label"
        htmlFor="expenseLabel"
      >
        expenseForm.expenseLabel
      </label>
      <div
        className="form-field__input"
      >
        <input
          data-testid="expenseLabel-test"
          id="expenseLabel"
          name="expenseLabel"
          onChange={[Function]}
          type="expenseLabel"
          value=""
        />
      </div>
    </div>
    <div
      className="form-field"
    >
      <label
        className="form-field__label"
        htmlFor="amount"
      >
        expenseForm.amountLabel
      </label>
      <div
        className="form-field__input"
      >
        <input
          data-testid="amount-test"
          id="amount"
          name="amount"
          onChange={[Function]}
          type="number"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="form-submit"
  >
    <button
      data-testid="reset-button-test"
      type="reset"
    >
      expenseForm.resetButton
    </button>
    <button
      data-testid="submit-button-test"
      type="submit"
    >
      expenseForm.submitButton
    </button>
  </div>
</form>
`;
